src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,661 @@
+    size_t vertices = boost::num_vertices(pattern);
+    size_t max_edges = vertices * (vertices + 1) / 2;
+    if (boost::num_edges(pattern) != max_edges) {
+        std::vector<size_t> indices(vertices);
+        std::iota(std::begin(indices), std::end(indices), 0);
+
+        std::vector<std::vector<size_t>> pairs = GetSubsets<size_t>(indices, 2);
+        for (size_t j = 0; j < vertices; ++j) {
+            std::vector<size_t> element = {j, j};
+            pairs.push_back(std::move(element));
+        }
+        vertex_t origin, finish;
+        for (auto& pair : pairs) {
+            vertex_t u = boost::vertex(pair.at(0), pattern);
+            vertex_t v = boost::vertex(pair.at(1), pattern);
+
+            if (!boost::edge(u, v, pattern).second) {
+                origin = u;
+                finish = v;
+                break;
+            }
+        }
+
+        for (auto& label : edge_labels) {
+            graph_t new_pattern = {};
+            boost::copy_graph(pattern, new_pattern);
+            boost::add_edge(origin, finish, {std::move(label)}, new_pattern);
+
+            bool iso_exists = false;
+            UpdateExisted(iso_exists, new_pattern, new_patterns, new_forbidden_rules_set,
+                          forbidden_rules);
+            if (iso_exists) {
+                continue;
+            }
+
+            Embeddings new_embeddings = {};
+            for (auto& embedding : embeddings) {
+                if (boost::edge(embedding.at(origin), embedding.at(finish), graph).second) {
+                    new_embeddings.push_back(std::move(embedding));
+                }
+            }
+            if (!new_embeddings.empty()) {
+                new_forbidden_rules_set.push_back(std::move(forbidden_rules));
+                new_patterns.push_back(std::move(new_pattern));
+                new_embeddings_set.push_back(std::move(new_embeddings));
+            }
+        }
+    }
+}
+
+void AddVertex(std::set<std::string> const& vertex_labels, std::set<std::string> const& edge_labels,
+               graph_t const& graph, graph_t const& pattern, Embeddings const& embeddings,
+               Rules const& forbidden_rules, std::vector<graph_t>& new_patterns,
+               std::vector<Embeddings>& new_embeddings_set,
+               std::vector<Rules>& new_forbidden_rules_set) {
+    for (auto const& edge_label : edge_labels) {
+        for (auto const& vertex_label : vertex_labels) {
+            for (size_t j = 0; j < boost::num_vertices(pattern); ++j) {
+                vertex_t u = boost::vertex(j, pattern);
+                Embeddings new_embeddings = {};
+
+                graph_t new_pattern = {};
+                boost::copy_graph(pattern, new_pattern);
+                std::map<std::string, std::string> attributes = {{"label", vertex_label}};
+                vertex_t new_vertex = boost::add_vertex(
+                        Vertex{static_cast<int>(boost::num_vertices(pattern)), attributes},
```

Comment:
The `node_id` field of the `Vertex` class has the `int` type
The code of three merged algorithms depends on this class, if you change the field type `node_id` to `size_t`, you will also have to refactor a large amount of code. If this is necessary, it will be done, just decided to clarify

Reply:
it is contained in the graph_descriptor.h file in the merged code